### PR TITLE
Resolve CVE-2015-9284

### PR DESCRIPTION
**Pull request by Hamms**
Turns out there's a fairly obscure set of conditions which can result in oauths that use GET requests being insecure. To prevent that, we install a gem which proactively prevents that request method from being used, and update the one part of our system which was using GET requests.

That part of our system is the "manage linked accounts" form on the user settings page; it implements the ability to distinguish between a "connect" oauth action and a "sign in/up" action by setting a timed session variable and redirecting to the oauth endpoint. Because we can no longer redirect, the new version distinguishes between different oauth actions by using oauth parameters, specifically by adding a new 'action' parameter. (Note that we could also in the future use this parameter to distinguish between sign ins and sign ups!)

Because the new connection pathway requires us to use form posts and redirects rather than clientside actions, we also update the disconnection pathway to use the same for parallelism.

**Reviewer note:** the rails unit test updates will probably be easiest to review [with whitespace changes ignored](https://github.com/code-dot-org/code-dot-org/pull/35425/commits/0b3b053306a1bca9575cd64e9121ea6bdceec4b7?w=1), since the removal of the timed session variable means we can take a lot of stuff out of Timecop blocks

## Links

- [CVE report](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-9284)
- [Resolution how-to](https://github.com/omniauth/omniauth/wiki/Resolving-CVE-2015-9284)

## Testing story

No new tests were added; I played around with adding some initially, but eventually realized that the existing tests in this space are already more than sufficient, so I merely updated those.

## Follow-up

As a result of removing the clientside `disconnect` logic, there's a bunch of redux work that is no longer used. I'll take that out in a follow-up PR, in the interests of keeping this one focused.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -85,10 +85,16 @@ def all
 
   # Call GET /users/auth/:provider/connect and the callback will trigger this code path
   def connect_provider
-    return head(:bad_request) unless can_connect_provider?
+    unless current_user&.migrated?
+      flash.alert = I18n.t('auth.migration_required')
+      return redirect_to edit_user_registration_path
+    end
```
Comment (islemaster) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb:91]: Nice, this is definitely a UX improvement.

**Review by islemaster — APPROVED** [review]
:tada: Great fix.  In the end this feels simpler, as well.

**Comment by Hamms** [comment]
I agree! It is a bummer to lose the nice no-reload-required disconnect experience, but there is a satisfaction to getting back to basics with good ol' form requests.

**Review by maddiedierker — APPROVED** [review]
LGTM! +1 to brad's comment -- thank you for simplifying the error handling UX here as well 🎉

**Comment by Hamms** [comment]
Oops; I accidentally pushed that "remove clientside disconnect" functionality I was intending to do in a follow-up. Reverted that commit, will reapply it in https://github.com/code-dot-org/code-dot-org/pull/35576

**Pull request merged**